File: scripts/scenarios/scenarios/01_account_lifecycle.py
"""Scenario 1: "First Day on the Network" — Account Lifecycle & Identity

Luna Starfield creates her account on the PDS, her DID gets registered
in PLC, she sets up her profile, and verifies her identity resolves
correctly through PLC.

Services: PDS, PLC
"""

from __future__ import annotations

import sys
import time
from pathlib import Path

# Add repo root to path
_project_root = str(Path(__file__).resolve().parent.parent.parent.parent)
if _project_root not in sys.path:
    sys.path.insert(0, _project_root)

from scripts.lib.atproto import XrpcClient, get_character, PDS1, ScenarioResult, timed_call


def run() -> ScenarioResult:
    result = ScenarioResult("Account Lifecycle & Identity")
    result.start()

    pds = XrpcClient(PDS1)
    luna = get_character("luna")

    timed_call(result, "Server health check",
               lambda: pds.wait_for_healthy(timeout=30))
    if result.failed > 0:
        result.finish()
        return result

    desc = timed_call(
        result, "Describe server",
        lambda: pds.accounts.describe_server(),
        detail_fn=lambda d: f"domains={d.get('availableUserDomains')}",
    )

    session = timed_call(
        result, "Create account",
        lambda: pds.accounts.create_account(luna.handle, luna.email, luna.password),
        detail_fn=lambda s: f"did={s['did']}",
    )
    if session:
        luna.did = session["did"]
        luna.access_jwt = session["accessJwt"]
        luna.refresh_jwt = session.get("refreshJwt")
    else:
        result.finish()
        return result

    timed_call(
        result, "Get session",
        lambda: pds.accounts.get_session(luna.access_jwt),
        detail_fn=lambda s: f"did={s.get('did')}",
    )

    timed_call(
        result, "Resolve handle",
        lambda: pds.identity.resolve_handle(luna.handle),
        detail_fn=lambda r: f"did={r.get('did')}",
    )

    try:
        import requests
        plc_resp = requests.get(f"http://localhost:2582/{luna.did}", timeout=10)
        if plc_resp.status_code == 200:
            did_doc = plc_resp.json()
            did_field = did_doc.get("id") or did_doc.get("did")
            assert did_field == luna.did, f"PLC DID mismatch: expected {luna.did}, got {did_field}"
            result.step_passed("PLC DID resolution", f"method={did_doc.get('verificationMethod', 'N/A')}")
        else:
            result.step_skipped("PLC DID resolution", f"PLC returned {plc_resp.status_code}")
    except Exception as exc:
        result.step_skipped("PLC DID resolution", str(exc))

    profile = {
        "$type": "app.bsky.actor.profile",
        "displayName": "Luna Starfield",
        "description": "Astronomy enthusiast. Looking up, always.",
    }
    timed_call(
        result, "Create profile",
        lambda: pds.records.create_record(luna.did, "app.bsky.actor.profile", profile, luna.access_jwt),
        detail_fn=lambda r: f"uri={r['uri']}",
    )

    timed_call(
        result, "Get profile",
        lambda: pds.feed.get_profile(luna.did, token=luna.access_jwt),
        detail_fn=lambda p: f"displayName={p.get('displayName')}",
    )

    if luna.refresh_jwt:
        refreshed = timed_call(
            result, "Refresh session",
            lambda: pds.accounts.refresh_session(luna.refresh_jwt),
            detail_fn=lambda r: f"accessJwt={r['accessJwt'][:20]}...",
        )
        if refreshed:
            luna.access_jwt = refreshed["accessJwt"]
    else:
        result.step_skipped("Refresh session", "No refreshJwt available")

    timed_call(
        result, "Invalid login rejected",
        lambda: pds.accounts.create_session(luna.handle, "wrong_password"),
        expect_failure=True,
    )

    try:
        pds.accounts.delete_session(luna.access_jwt)
        result.step_passed("Delete session (logout)")
    except Exception as exc:
        result.step_skipped("Delete session", str(exc))

    result.finish()
    return result


if __name__ == "__main__":
    result = run()
    result.print_summary()
    sys.exit(result.exit_code)

File: scripts/scenarios/scenarios/01_account_lifecycle.ts
/**
 * @module scenarios/01_account_lifecycle
 *
 * Scenario: Account Lifecycle & Identity verification.
 *
 * Behavior:
 * - Creates a new user account on the PDS.
 * - Verifies session establishment, handle resolution, and PLC DID resolution.
 * - Creates and retrieves a user profile.
 * - Tests session refreshing and invalid login handling.
 * - Logs the user out.
 *
 * Expectations:
 * - All lifecycle operations complete successfully with valid responses.
 */

import { XrpcClient } from "../../lib/deno/client.ts";
import { getActor, PDS1, SERVICE_URLS } from "../../lib/deno/config.ts";
import { createAccountOrLogin, ScenarioResult, timedCall } from "../../lib/deno/runner.ts";
export { ScenarioResult, StepResult, StepStatus } from "../../lib/deno/runner.ts";
export type { ScenarioReport } from "../../lib/deno/runner.ts";
import { assert } from "../../lib/deno/assertions.ts";

/**
 * Executes the scenario logic.
 * @returns A promise that resolves to the scenario result
 */
export async function run(): Promise<ScenarioResult> {
  const result = new ScenarioResult("Account Lifecycle & Identity");
  result.start();

  const pds = new XrpcClient(PDS1);
  const luna = getActor("luna");

  await timedCall(
    result,
    "Server health check",
    async () => {
      await pds.raw.xrpcGet("com.atproto.server.describeServer");
    },
  );

  if (result.failed > 0) {
    result.finish();
    return result;
  }

  await timedCall(
    result,
    "Describe server",
    async () => {
      return await pds.raw.get("com.atproto.server.describeServer");
    },
    (d) => `domains=${d.availableUserDomains}`,
  );

  const session = await timedCall(
    result,
    "Create account",
    () => createAccountOrLogin(pds, luna),
    (s) => `did=${s.did}`,
  );

  if (session) {
    luna.did = session.did;
    luna.accessJwt = session.accessJwt;
    luna.refreshJwt = session.refreshJwt;
  } else {
    result.finish();
    return result;
  }

  await timedCall(
    result,
    "Get session",
    async () => {
      const res = await pds.as(luna).api.com.atproto.server.getSession();
      return res;
    },
    (s) => `did=${s.did}`,
  );

  await timedCall(
    result,
    "Resolve handle",
    async () => {
      const res = await pds.api.com.atproto.identity.resolveHandle({ handle: luna.handle });
      return res;
    },
    (r) => `did=${r.did}`,
  );

  try {
    const plcResp = await pds.raw.httpGet(`${SERVICE_URLS.plc}/${luna.did}`);
    const didDoc = plcResp;
    const didField = didDoc.id || didDoc.did;
      assert.equal(didField, luna.did, `PLC DID mismatch: expected ${luna.did}, got ${didField}`);
      result.stepPassed(
        "PLC DID resolution",
        `method=${didDoc.verificationMethod ? "present" : "N/A"}`,
      );
  } catch (exc: any) {
    result.stepSkipped("PLC DID resolution", exc.message || String(exc));
  }

  const profile = {
    $type: "app.bsky.actor.profile",
    displayName: "Luna Starfield",
    description: "Astronomy enthusiast. Looking up, always.",
  };

  await timedCall(
    result,
    "Create profile",
    async () => {
      const res = await pds.as(luna).repo.createRecord({
        collection: "app.bsky.actor.profile",
        record: profile,
      });
      return res;
    },
    (r) => `uri=${r.uri}`,
  );

  await timedCall(
    result,
    "Get profile",
    async () => {
      const res = await pds.as(luna).raw.get(
        "app.bsky.actor.getProfile",
        { actor: luna.did },
      );
      return res;
    },
    (p) => `displayName=${p?.displayName}`,
  );

  if (luna.refreshJwt) {
    const refreshed = await timedCall(
      result,
      "Refresh session",
      async () => {
        const res = await pds.as({ accessJwt: luna.refreshJwt }).api.com.atproto.server.refreshSession();
        return res;
      },
      (r) => `accessJwt=${r.accessJwt.substring(0, 20)}...`,
    );
    if (refreshed) {
      luna.accessJwt = refreshed.accessJwt;
      luna.refreshJwt = refreshed.refreshJwt;
    }
  } else {
    result.stepSkipped("Refresh session", "No refreshJwt available");
  }

  await timedCall(
    result,
    "Invalid login rejected",
    async () => {
      await pds.raw.xrpcPost("com.atproto.server.createSession", { identifier: luna.handle, password: "wrong_password" });
    },
    undefined,
    true,
  );

  try {
    await pds.as(luna).api.com.atproto.server.deleteSession();
    result.stepPassed("Delete session (logout)");
  } catch (exc: any) {
    result.stepSkipped("Delete session", exc.message || String(exc));
  }

  result.finish();
  return result;
}

if (import.meta.main) {
  run().then((res) => {
    console.log(res.summary());
    Deno.exit(res.ok ? 0 : 1);
  });
}
